src/primitive_crypto.cc

```diff
@@ -118,6 +120,206 @@ PRIMITIVE(siphash_get) {
+  if (err != 0) {
+    group->unregister_resource(gcm_context);
+    return tls_error(null, process, err);
+  }
+  
+  proxy->set_external_address(gcm_context);
+  return proxy;
+}
+
+PRIMITIVE(gcm_close) {
+  ARGS(GcmContext, context);
+  context->resource_group()->unregister_resource(context);
+  context_proxy->clear_external_address();
+  return process->program()->null_object();
+}
+
+// Start the encryption of a message, or TLS record.  Takes a 12 byte nonce.
+// As described in RFC5288 section 3, the first 4 bytes of the nonce are kept
+// secret, and the next 8 bytes are transmitted along with each record.
+// Internally these primitives will add 4 more bytes of block counter, starting
+// at 1, to form a 16 byte IV.  It is vital that the nonce is not reused, and
+// this is normally achieved by counting up the explicit part by one for each
+// record that is encrypted.  In TLS this means that this part of the nonce
+// corresponds to the sequence number of the record.
+PRIMITIVE(gcm_start_message) {
+  ARGS(GcmContext, context, int, length, Blob, authenticated_data, Blob, nonce);
+  if (context->remaining_length_in_current_message() != 0) INVALID_ARGUMENT;
+  context->set_remaining_length_in_current_message(length);
+  if (nonce.length() != GcmContext::NONCE_SIZE) INVALID_ARGUMENT;
+  int mode = context->is_encrypt() ? MBEDTLS_GCM_ENCRYPT : MBEDTLS_GCM_DECRYPT;
+  int result = mbedtls_gcm_starts(
+      context->gcm_context(),
+      mode,
+      nonce.address(),
+      nonce.length(),
+      authenticated_data.address(),  // No additional data.
+      authenticated_data.length());
+  if (result != 0) return tls_error(null, process, result);
+
+  return process->program()->null_object();
+}
+
+/**
+If the result byte array was big enough, returns a Smi to indicate how much
+  data was placed in it.
+If the result byte array was not big enough, returns null.  In this case no
+  data was consumed.
+*/
+PRIMITIVE(gcm_add) {
+  ARGS(GcmContext, context, Blob, data, MutableBlob, result);
+  const uint8* data_address = data.address();
+  int data_length = data.length();
+  if (data_length == 0) return Smi::from(0);
+  int remains = context->remaining_length_in_current_message();
+  if (remains < data_length) OUT_OF_BOUNDS;
+  remains -= data_length;
+
+  // Start by copying into the temporary buffer in the context.
+  const int to_copy = Utils::min(
+      GcmContext::BLOCK_SIZE - context->buffered_bytes(),
+      data_length);
+  memcpy(context->buffered_data() + context->buffered_bytes(), data_address, to_copy);
+  data_address += to_copy;
+  data_length -= to_copy;
+  const int buffered = context->buffered_bytes() + to_copy;
+  if (buffered < GcmContext::BLOCK_SIZE && remains != 0) {
+    // Success.  We copied all the data into the internal buffer, and so the
+    // output byte array is inevitably big enough.
+    // Update the context and the number of bytes of new output.
```

Comment:
```suggestion
    // Success.  We copied all the data into the internal buffer.
    // Since we haven't filled up a full block yet, we can't run the crypt function yet
    // Update the context and let the caller know that we successfully handled the data.
```